### PR TITLE
storage: Support vector index and ANN hint

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9032

Problem Summary:

### What is changed and how it works?

Pick https://github.com/tidbcloud/tiflash-cse/pull/156, https://github.com/tidbcloud/tiflash-cse/pull/162, https://github.com/tidbcloud/tiflash-cse/pull/163, https://github.com/tidbcloud/tiflash-cse/pull/164

```commit-message

```
Changes:
* Support one or no vector index on the vector column
* Enable `ExtendColumnStat` in DMFile meta for storing the meta of vector index
* Write: When vector index exist, `DMFileWriter::addStreams` will generate the index while writing down a new DMFile
  * class `VectorIndex` generates the index by HNSW algorithm
* Read: `DMFileBlockInputStreamBuilder` will try to generate a `DMFileWithVectorIndexBlockInputStream`. If vector index is unavailable, the read fallback to normal read and filter. In `DMFileWithVectorIndexBlockInputStream`, it
  * load vector index and apply ANN vector search algorithm, this will generate a rowid bitmap for later reading
  * load the vector column data (from vector index) and other column data from disk

TODO:
* Support building vector index in background pool
* Support building multiple vector index on the same vector column
* Support adding/dropping vector index following TiDB DDL
* Confirm the following logic
https://github.com/pingcap/tiflash/blob/9fab7b262b925774196efeb6e7677f8aa2f92a7e/dbms/src/Storages/DeltaMerge/File/DMFileBlockInputStream.cpp#L145-L147

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/TiDB/Schema/TiDB.cpp [hunk 1]
```diff
@@ -452,6 +466,27 @@ try
...
+        vector_index = std::make_shared<VectorIndexInfo>();
+
+        auto vector_kind = magic_enum::enum_cast<VectorIndexKind>(vector_index_json->getValue<String>("kind"));
+        RUNTIME_CHECK(vector_kind.has_value());
+        RUNTIME_CHECK(vector_kind.value() != VectorIndexKind::INVALID);
+        vector_index->kind = vector_kind.value();
+
+        vector_index->dimension = vector_index_json->getValue<UInt64>("dimension");
+        RUNTIME_CHECK(vector_index->dimension > 0);
+        RUNTIME_CHECK(vector_index->dimension <= 16000); // Just a protection
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/TiDB.cpp:482]:
```suggestion
        RUNTIME_CHECK(vector_index->dimension <= 16383); // Just a protection
```

This should be updated to 16383 for MySQL compatibility according to the feature spec https://github.com/pingcap/tidb/pull/55002#discussion_r1703687255

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterView.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    size_t count() const
+    {
+        size_t n = 0;
+        for (UInt32 i = 0; i < size(); i++)
+        {
+            if (get(i))
+                n++;
+        }
+        return n;
+    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterView.h:60]: Add `friend class BitmapFilterView;` to BitmapFilter and use this
```suggestion
    // Return how many valid rows.
    size_t count() const
    {
        return std::count(
            filter->filter.cbegin() + filter_offset,
            filter->filter.cbegin() + filter_offset + filter_size,
            true);
    }
```

**Review by breezewish — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/File/DMFileMetaV2.cpp [hunk 3]
```diff
@@ -177,16 +177,10 @@ void DMFileMetaV2::finalize(
...
+    std::array meta_block_handles = {
         writeSLPackStatToBuffer(tmp_buffer),
         writeSLPackPropertyToBuffer(tmp_buffer),
-#if 1
-        writeColumnStatToBuffer(tmp_buffer),
-#else
-        // ExtendColumnStat is not enabled yet because it cause downgrade compatibility, wait
-        // to be released with other binary format changes.
         writeExtendColumnStatToBuffer(tmp_buffer),
-#endif
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/File/DMFileMetaV2.cpp:189]: @JaySon-Huang Please confirm those changes.

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**